plenoptic/simulate/canonical_computations/steerable_pyramid_freq.py

```diff
@@ -321,89 +316,98 @@ def forward(self, x, scales=[]):
 
             if i in scales:
-
+                #high-pass mask is selected based on the current scale
                 himask = self._himasks[i]
+                #compute filter output at each orientation
                 for b in range(self.num_orientations):
+                    '''
+                    band pass filtering is done in the fourier space as multiplying by the fft of a gaussian derivative.
+                    The oriented dft is computed as a product of the fft of the low-passed component,
+                    the precomputed anglemask (specifies orientation), and the precomputed hipass mask (creating a bandpass filter)
+                    the complex_const variable comes from the Fourier transform of a gaussian derivative.
+                    Based on the order of the gaussian, this constant changes.
+                    
+                    '''
                     anglemask = self._anglemasks[i][b]
-
-                    # bandpass filtering
-                    banddft = lodft * anglemask * himask
-                    banddft = torch.unbind(banddft, -1)
-                    # (x+yi)(u+vi) = (xu-yv) + (xv+yu)i
                     complex_const = np.power(np.complex(0, -1), self.order)
-                    banddft_real = complex_const.real * banddft[0] - complex_const.imag * banddft[1]
-                    banddft_imag = complex_const.real * banddft[1] + complex_const.imag * banddft[0]
-                    # preallocation and then filling in is much more
-                    # efficient than using stack
-                    banddft = torch.empty((*banddft_real.shape, 2), device=banddft_real.device)
-                    banddft[..., 0] = banddft_real
-                    banddft[..., 1] = banddft_imag
-
-                    band = batch_ifftshift(banddft)
-                    band = torch.ifft(band, signal_ndim=2, normalized=self.fft_normalize)
+                    banddft = complex_const * lodft * anglemask * himask
+                    #fft output is then shifted to center frequencies
+                    band = fft.ifftshift(banddft)
+                    #ifft is applied to recover the filtered representation in spatial domain
+                    band = fft.ifft2(band, dim=(-2,-1), norm=self.fft_norm)
+                    
+                    #for real pyramid, take the real component of the complex band
                     if not self.is_complex:
-                        band = torch.unbind(band, -1)[0]
-                        self.pyr_coeffs[(i, b)] = band
-                        self.pyr_size[(i, b)] = tuple(band.shape[-2:])
+                        pyr_coeffs[(i, b)] = band.real
                     else:
+                        '''
+                        Because the input signal is real, to maintain a tight frame 
+                        if the complex pyramid is used, magnitudes need to be divided by sqrt(2) 
+                        because energy is doubled.
+                        '''
```

Comment:
can the pyramid only be tight frame if it's complex? (all the `if self.tight_frame` statements are nested within or paired with `if self.is_complex`). If so, we should clarify that in the docstring and raise an exception if they try to set `is_complex=False, tight_frame=True`

Reply:
The pyramid can be tight_frame even if it's not complex- see the new pyramid file for updated code- not all of the if statements are nested with a self.is_complex now.